torch/distributed/fsdp/flat_param.py

```diff
@@ -1398,7 +1409,16 @@ def _use_unsharded_grad_views(self) -> None:
                 f"{self.flat_param._fqns[i]} is missing",
             )
             param = getattr(module, param_name)
-            param.grad = view
+            if param.shape != view.shape:
+                # NOTE: This is a hack using `.data` to side step the
+                # check that parameter/gradient sizes match. Here,
+                # `param` has the sharded size; `grad` has the unsharded size.
+                # This happens when running in `no_sync()`.
+                if param.grad is None:
+                    param.grad = torch.empty_like(param)
```

Comment:
I assume this gonna fail if we create a zero-sized tensor here? And this is what you referred to as "the check that parameter/gradient sizes match"?

Reply:
We should not ever create a zero-sized tensor here because this is `_use_unsharded_grad_views()`, not `_use_sharded_grad_views()`.

The check I am referring to is that for any `param: nn.Parameter`, PyTorch checks that `param` and `param.grad` have the same shape if `param.grad is not None`. However, here, `param` may have the sharded size (e.g. `[5]`), while `param.grad may have the unsharded size (e.g.`[9]`). (The example numbers here show the padding case for 2 ranks and an unsharded size of 10 numel.)